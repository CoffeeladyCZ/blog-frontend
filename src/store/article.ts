import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Article } from '../types/Articles';

interface ArticleState {
  listArticles: Article[];
}

const initialState: ArticleState = {
  listArticles: []
};

export const articleSlice = createSlice({
  name: 'listArticles',
  initialState,
  reducers: {
    setListArticles: (state, action: PayloadAction<Article[]>) => {
      state.listArticles = action.payload;
    }
  }
});

export const { setListArticles } = articleSlice.actions;
export default articleSlice.reducer;
