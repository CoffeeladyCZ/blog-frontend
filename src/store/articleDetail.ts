import { ArticleDetailTypes } from './../types/Articles';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface ArticleDetailState {
  articleDetail: ArticleDetailTypes | null;
}

const initialArticleDetailState: ArticleDetailState = {
  articleDetail: null
};

export const articleDetailSlice = createSlice({
  name: 'articleDetail',
  initialState: initialArticleDetailState,
  reducers: {
    setArticleDetail: (state, action: PayloadAction<ArticleDetailTypes | null>) => {
      state.articleDetail = action.payload;
    }
  }
});

export const { setArticleDetail } = articleDetailSlice.actions;

export const articleDetailReducer = articleDetailSlice.reducer;
