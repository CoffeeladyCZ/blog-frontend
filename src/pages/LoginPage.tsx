import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import axios from 'axios';
import { useTranslation } from 'react-i18next';
import { Navigate } from 'react-router-dom';
import Cookies from 'js-cookie';

import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../store/store';
import { setLogin } from '../store/login';

import { Button, Card, Grid, TextField, Typography, CardContent, CardActions } from '@mui/material';
import { styled } from '@mui/material/styles';

type FormValues = {
  username: string;
  password: string;
};

const StyledTypography = styled(Typography)`
  font-size: 1.75rem;
  margin-bottom: 1.5rem;
`;

const StyledCard = styled(Card)`
  width: 368px;
  margin: 0 auto;
  margin-top: 50px;
  box-shadow: "0px 16px 48px rgba(0, 0, 0, 0.175);
`;

const StyledCardActions = styled(CardActions)`
  padding: 16px;
  justify-content: end;
`;

const apiLogin = 'https://fullstack.exercise.applifting.cz/login';

const LoginPage: React.FC = () => {
  const { t } = useTranslation();
  const [errorMessage, setErrorMessage] = useState('');
  const [isLoading, setIsLoading] = useState(false);

  const dispatch = useDispatch();
  const login = useSelector((state: RootState) => state.login.login);

  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<FormValues>({
    mode: 'onChange'
  });

  const config = {
    headers: {
      'X-API-KEY': 'd4234190-5f6b-4d51-b3f3-80cc4810d0b7'
    }
  };

  const onSubmit = async (data: FormValues) => {
    setIsLoading(true);

    console.log('data', data);
    let response;
    let access_token;
    try {
      response = await axios.post(apiLogin, data, config);
      access_token = await response.data.access_token;
      Cookies.set('token', access_token);
      dispatch(setLogin(true));
    } catch (error) {
      console.error(error);
      setIsLoading(false);
      setErrorMessage((error as any)?.response?.data?.message);
    }
    console.log('token', access_token);
  };

  return login ? (
    <Navigate to="/articles" replace />
  ) : (
    <StyledCard>
      <form onSubmit={handleSubmit(onSubmit)}>
        <CardContent>
          <StyledTypography variant="h3">Log In</StyledTypography>
          <Grid container spacing={2}>
            <Grid item xs={12}>
              <TextField
                label="E-mail"
                {...register('username', {
                  required: t('errorMessage:email')
                })}
                error={Boolean(errors.username)}
                helperText={errors.username?.message}
                id="username"
                placeholder="Login"
                size="small"
                fullWidth
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                label="Password"
                type="password"
                {...register('password', {
                  required: t('errorMessage:password')
                })}
                error={Boolean(errors.password)}
                helperText={errors.password?.message}
                id="password"
                placeholder="********"
                size="small"
                fullWidth
              />
            </Grid>
          </Grid>
          {errorMessage && (
            <Typography variant="h6" sx={{ color: '#d32f2f' }} id="error">
              {errorMessage}
            </Typography>
          )}
        </CardContent>
        <StyledCardActions>
          <Button
            size="small"
            color="primary"
            type="submit"
            disabled={isLoading}
            variant="contained">
            Log In
          </Button>
        </StyledCardActions>
      </form>
    </StyledCard>
  );
};

export default LoginPage;
