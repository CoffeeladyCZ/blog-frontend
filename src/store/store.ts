import { configureStore } from '@reduxjs/toolkit';
import { articleSlice } from './article';
import { articleDetailSlice } from './articleDetail';
import { loginSlice } from './login';

const store = configureStore({
  reducer: {
    articleList: articleSlice.reducer,
    articleDetail: articleDetailSlice.reducer,
    login: loginSlice.reducer
  }
});

export type AppStore = ReturnType<typeof store.getState>;
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
