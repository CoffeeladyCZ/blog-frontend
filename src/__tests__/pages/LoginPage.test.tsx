import { render, screen, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { setLogin } from '../../store/login';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { I18nextProvider } from 'react-i18next';
import i18next from '../../i18n/config';

import LoginPage from '../../pages/LoginPage';

describe('LoginPage', () => {
  let axiosMock: MockAdapter;
  const mockStore = configureStore([]);
  const initialState = {
    login: {
      login: false
    }
  };

  beforeEach(async () => {
    jest.spyOn(console, 'error');
    axiosMock = new MockAdapter(axios);
    const store = mockStore(initialState);

    render(
      <I18nextProvider i18n={i18next}>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/login']}>
            <LoginPage />
          </MemoryRouter>
        </Provider>
      </I18nextProvider>
    );
  });

  test('should snapshot with data', () => {
    act(() => {
      const store = mockStore(initialState);

      const { asFragment } = render(
        <I18nextProvider i18n={i18next}>
          <Provider store={store}>
            <MemoryRouter initialEntries={['/login']}>
              <LoginPage />
            </MemoryRouter>
          </Provider>
        </I18nextProvider>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  test('renders component', async () => {
    const loginEmail = await screen.getByTestId('loginEmail');
    expect(loginEmail).toBeInTheDocument();
    const loginPassword = await screen.getByTestId('loginPassword');
    expect(loginPassword).toBeInTheDocument();
    const loginButton = await screen.getByTestId('loginButton');
    expect(loginButton).toBeInTheDocument();
  });

  test('handles onSubmit correctly - success is false', async () => {
    const store = mockStore(initialState);

    const mockLoginResponse = {
      success: false,
      error: {
        response: {
          data: {
            message: 'Login failed'
          }
        }
      }
    };

    await axiosMock.onPost('/login').reply(400, mockLoginResponse);

    const emailInput = screen.getByTestId('loginEmail');
    userEvent.type(emailInput, 'test@example.com');

    const passwordInput = screen.getByTestId('loginPassword');
    userEvent.type(passwordInput, 'password');

    const submitButton = screen.getByTestId('loginButton');
    fireEvent.click(submitButton);

    expect(store.getActions()).toEqual([]);
  });

  test('handles onSubmit correctly - success is true', async () => {
    const store = mockStore(initialState);
    const mockLoginResponse = {
      success: true
    };

    await axiosMock.onPost('/login').reply(200, mockLoginResponse);

    const emailInput = screen.getByTestId('loginEmail');
    userEvent.type(emailInput, 'test@example.com');

    const passwordInput = screen.getByTestId('loginPassword');
    userEvent.type(passwordInput, 'password123');

    const submitButton = screen.getByTestId('loginButton');
    fireEvent.click(submitButton);

    store.dispatch(setLogin(true));

    expect(store.getActions()).toEqual([{ type: 'login/setLogin', payload: true }]);
  });
});
