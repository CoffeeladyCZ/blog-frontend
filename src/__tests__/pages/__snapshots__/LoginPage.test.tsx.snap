// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should snapshot with data 1`] = `<DocumentFragment />`;
